sdk/iptables/iptables.go

```diff
@@ -0,0 +1,158 @@
+package iptables
+
+import (
+	"errors"
+	"fmt"
+	"os/exec"
+	"strconv"
+)
+
+const (
+	// Chain to intercept inbound traffic
+	ProxyInboundChain = "PROXY_INBOUND"
+
+	// Chain to redirect inbound traffic to the proxy
+	ProxyInboundRedirectChain = "PROXY_IN_REDIRECT"
+
+	// Chain to intercept outbound traffic
+	ProxyOutputChain = "PROXY_OUTPUT"
+
+	// Chain to redirect outbound traffic to the proxy
+	ProxyOutputRedirectChain = "PROXY_REDIRECT"
+)
+
+// Config is used to configure which traffic interception and redirection
+// rules should be applied with the iptables commands.
+type Config struct {
+	// ProxyUserID is the user ID of the proxy process.
+	ProxyUserID string
+
+	// ProxyInboundPort is the port of the proxy's inbound listener.
+	ProxyInboundPort int
+
+	// ProxyInboundPort is the port of the proxy's outbound listener.
+	ProxyOutboundPort int
+
+	// IptablesProvider is the Provider that will apply iptables rules.
+	IptablesProvider Provider
+}
+
+// Provider is an interface for executing iptables rules.
+type Provider interface {
+	// AddRule adds a rule without executing it.
+	AddRule(name string, args ...string)
+	// ApplyRules executes rules that have been added via AddRule.
+	ApplyRules() error
+	// Rules returns the list of rules that have been added but not applied yet.
+	Rules() []string
+}
+
+// iptablesExecutor implements IptablesProvider using exec.Cmd.
+type iptablesExecutor struct {
+	commands []*exec.Cmd
+}
+
+func (i *iptablesExecutor) AddRule(name string, args ...string) {
+	i.commands = append(i.commands, exec.Command(name, args...))
+}
+
+func (i *iptablesExecutor) ApplyRules() error {
+	_, err := exec.LookPath("iptables")
+	if err != nil {
+		return err
+	}
+
+	for _, cmd := range i.commands {
+		err := cmd.Run()
+		if err != nil {
+			output, err := cmd.CombinedOutput()
+
+			if err != nil {
+				return fmt.Errorf("failed to run command: %s, err: %v", cmd.String(), err)
+			}
+
+			return fmt.Errorf("failed to run command: %s, err: %v, output: %s", cmd.String(), err, string(output))
+		}
+	}
+
+	return nil
+}
+
+func (i *iptablesExecutor) Rules() []string {
+	var rules []string
+	for _, cmd := range i.commands {
+		rules = append(rules, cmd.String())
+	}
+
+	return rules
+}
+
+// Setup will set up iptables interception and redirection rules
+// based on the configuration provided in cfg.
+// This implementation was inspired by
+// https://github.com/openservicemesh/osm/blob/650a1a1dcf081ae90825f3b5dba6f30a0e532725/pkg/injector/iptables.go
+func Setup(cfg Config) error {
```

Comment:
Is this idempotent? What happens if it's run twice? Either with the same or different Config.

Should we attempt to drop the chains before adding them?

Reply:
Yeah, that's a great call-out. I've been thinking about your comment above about documenting that `ApplyRules` is not atomic, and I think it makes sense to make that one atomic and this one idempotent. I'm thinking to address it in a follow-up PR because most likely I'll switch to using `iptables-save` and `iptables-restore` to do that.